### ISSUE TITLE
Trying to access array offset on value of type null Product.php

### ISSUE THREAD

**Issue by kts1974** [PrestaShop/PrestaShop#34199]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add attachments

In the Product.php file an error occurs on the following lines:
        array_unshift($result, null);
        foreach ($result as &$value) {
            $value = $value['id_product'];
        }

This code adds null to the beginning of $result and in the first interaction it gives an error when querying the id_product variable because the first value of $result is null

This causes a php error:
Got error 'PHP message: [PHP Warning #2] Trying to access array offset on value of type null .../Product.php..., line 7245 

### Expected behavior

We hope it doesn't cause that any error or warning 

### Steps to reproduce

 Execute this php snippet:

// Enter your code here, enjoy!
        $result=array(array('id_product'=>10));
        array_unshift($result, null);
        foreach ($result as &$value) {
            $value = $value['id_product'];
	    print($value);
        }

You will get this: 
Warning: Trying to access array offset on value of type null in ...

Solution proposal:
       // Move array_unshift to the end
        foreach ($result as &$value) {
            $value = $value['id_product'];
	    print($value);
        }
       **array_unshift($result, null);**

### PrestaShop version(s) where the bug happened

8.1.2

### PHP version(s) where the bug happened

8.1.24

**Comment by florine2623** [comment]
Hello @kts1974 ,

Could you be more specific about the steps to reproduce the issue ?
Please share a screen record of your issue.

Thanks

**Comment by kts1974** [comment]
To reproduce the error you just have to run this little php code:

$result=array(array('id_product'=>10));
array_unshift($result, null);
foreach ($result as &$value) {
$value = $value['id_product'];
}

<img width="1224" alt="Screenshot" src="https://github.com/PrestaShop/PrestaShop/assets/147228594/6b6f1f7f-1a1b-42f2-ad5d-d1789aae49d0">


In any case, from the code in Prestashop you can see that it first adds null to the beginning of the array:
array_unshift($result, null);
and then try in the loop to read $value['id_product'];

As the first element is null, it gives an error when trying to obtain $value['id_product'];

**Comment by paulnoelcholot** [comment]
Hello,

Thanks for reporting this issue!
Ping @PrestaShop/prestashop-core-developers : Could someone please try to reproduce the issue, it's too technical!

Thanks in advance!

**Comment by kts1974** [comment]
This code also causes the following error when using the Prestashop webservice and modifying any product data:
[PHP Warning #2] Trying to access array offset on value of type null (.../httpdocs/classes/Product.php, line 7245

To understand the error you just have to analyze the Prestashop code:
array_unshift($result, null);
foreach ($result as &$value) {
$value = $value['id_product'];
}
In the first iteration of the loop it will try to access the key 'id_product' of the array, but it gives the error because the first value is null

In case it helps someone, while the error is studied and resolved, I have temporarily solved it by changing this code to:
foreach ($result as &$value) {
$value = $value['id_product'];
}
array_unshift($result, null);

**Comment by haveyougotanypets** [comment]
I have also found the same issue as @kts1974. We are unable to update any of our stock via the web service API.

This error seems to have arrived in 8.1.2. version 8.1.1 does not show this behaviour.

**Comment by kts1974** [comment]
While the possible bug is resolved in the official version, you can temporarily  change the following lines of code in   /classes/Product.php file (line 7245):

array_unshift($result, null);
foreach ($result as &$value) {
$value = $value['id_product'];
}

for this: code:
foreach ($result as &$value) {
$value = $value['id_product'];
}
array_unshift($result, null);

With this change you will be able to use the webservice without issues

**Comment by haveyougotanypets** [comment]
Thanks @kts1974 I followed your advise on Friday, all seems to be working now 👍

**Comment by HPODEV** [comment]
please, update! i have same error and the patch working

**Comment by KinG-InFeT** [comment]
hi, please merge the patch into main branch https://github.com/PrestaShop/PrestaShop/pull/36492

**Comment by matks** [comment]
hello I can see you are reporting a *defect* in class product.php but we need more information

please submit a list of user-driven steps we can follow like "Browse page (...) then submit payload (...) to API"

But a defect is not a bug report. We do not have the resources to transform a known defect into a bugfix. We are already aware of dozen of defects into the software, but it takes time to explore them, see how they can be fixed, test the fix is valid. This is why we can do nothing with a bug report that just mentions "I saw defect X in file Y".

We need a list of steps we can follow to reproduce the issue

`To reproduce the error you just have to run this little php code` or `Execute this php snippet` is not a valid reproduction ; we need a valid usecase **using prestashop**

**Comment by kts1974** [comment]
Hi matks 

Is not a defect is a bug. If the bug is not solved, the webservice service will not be able to be used for many product-related queries beacuse you will receive a php error.

To reproduce the bug you have to use the webservice api from a python code and retrieve the details of a order. Then you will get the next error:
Got error 'PHP message: [PHP Warning https://github.com/PrestaShop/PrestaShop/pull/2] Trying to access array offset on value of type null .../Product.php..., line 7245

The previous messages explain why the error occurs and how to solve the bug.

Regards

**Comment by jakubvondrak** [comment]
Same issue here - cannot update product via webservice. This however manifests only on certain PHP runtimes/configurations:

Environments I tested:
- OK: Production eshop PS 8.1.7, PHP 7.4.33
- ERROR: Production eshop PS 8.1.7, PHP 8.1.29
- ERROR: Production eshop PS 8.1.7, PHP 8.2.22
- ERROR: Production eshop PS 8.1.7, PHP 8.3.10
- OK: Docker image prestashop/prestashop:8.1.7 with demo products, PHP 8.1.29

So it is not because of PHP 8.1 but rather its configuration. I tried experiment with couple of options, but no luck figuring out which setting is causing it.

Reproduction steps:

1. Enable WS and add key E7PYVMVU3CAIUZCYKCGAR6IRDLRLY796 with all permissions on "products"
2. GET product: `curl -o product1.xml --user E7PYVMVU3CAIUZCYKCGAR6IRDLRLY796: http://prestashop:8080/api/products/1`
3. Remove read-only tags: `xmlstarlet ed -L -d "//manufacturer_name" -d "//quantity" product1.xml`
4. PUTTING product back results in HTTP 500: `curl -X PUT -H "Content-Type: application/xml" -T product1.xml --user E7PYVMVU3CAIUZCYKCGAR6IRDLRLY796: http://prestashop:8080/api/products/1`

```
<?xml version="1.0" encoding="UTF-8"?>
<prestashop xmlns:xlink="http://www.w3.org/1999/xlink">
<errors>
<error>
<code><![CDATA[3]]></code>
<message><![CDATA[[PHP Warning #2] Trying to access array offset on value of type null (.../www/classes/Product.php, line 7269)]]></message>
</error>
</errors>
</prestashop>
```

The error_log file also contains the warning (which isn't there when it works on PHP 7.4.33):
```
[08-Aug-2024 11:01:50 Europe/Prague] [PHP Warning #2] Trying to access array offset on value of type null (.../www/classes/Product.php, line 7269)
```

**Comment by KinG-InFeT** [comment]
it's absurd that a project like this is not maintained properly, I will abandon prestashop in every project.
PS: also the PR closed https://github.com/PrestaShop/PrestaShop/pull/36492 , shameful!

**Comment by eveneme** [comment]
+1

**Comment by matks** [comment]
And it's also absurd that in 2024, we still have people who do not understand the nature of open source 😞 .

If you want to understand what I mean, read https://snarky.ca/the-social-contract-of-open-source/ .

I will close and lock this issue because it does not comply with our [Code of Conduct](https://github.com/PrestaShop/PrestaShop/blob/09d5db04d3975a79eb6b02e674404a43123684e4/CODE_OF_CONDUCT.md).

I will nonetheless verify the steps provided by @jakubvondrak to see if I can help PR https://github.com/PrestaShop/PrestaShop/pull/36492 move on.

**Comment by matks** [comment]
@jakubvondrak Unfortunately I was not able to reproduce the issue. I think the steps you submitted are good and well described, however as you said it seems to depend on the PHP configuration.

On my localhost the following lines of code work well:
```
        array_unshift($result, null);
        foreach ($result as &$value) {
            $value = $value['id_product'];
        }
```

Even though I agree that
```
$result=array(array('id_product'=>10));
array_unshift($result, null);
foreach ($result as &$value) {
$value = $value['id_product'];
print($value);
}
```
Is supposed to throw a warning.

I can see the defect but I do not reproduce the buggy behavior reported.

I will submit a PR nonetheless to fix it, and do the needed items to allow it to move forward.

**Comment by matks** [comment]
PR submitted https://github.com/PrestaShop/PrestaShop/pull/36807

**Comment by florine2623** [comment]
Closing the issue as the bug is fixed with PR #36807
